src/bootstrap/src/core/build_steps/test.rs

```diff
@@ -2228,6 +2232,55 @@ impl BookTest {
         // Books often have feature-gated example text.
         rustbook_cmd.env("RUSTC_BOOTSTRAP", "1");
         rustbook_cmd.env("PATH", new_path).arg("test").arg(path);
+
+        // Books may also need to build dependencies. For example, `TheBook` has
+        // code samples which use the `trpl` crate. For the `rustdoc` invocation
+        // to find them them successfully, they need to be built first and their
+        // paths used to generate the
+        let libs = if !self.dependencies.is_empty() {
+            let mut lib_paths = vec![];
+            for dep in self.dependencies {
+                let mode = Mode::ToolBootstrap;
+                let target = builder.config.build;
+                let cargo = tool::prepare_tool_cargo(
+                    builder,
+                    self.compiler,
+                    mode,
+                    builder.config.build,
+                    Kind::Build,
+                    dep,
+                    SourceType::Submodule,
+                    &[],
+                );
+                let stamp = builder
+                    .cargo_out(self.compiler, mode, target)
+                    .join(PathBuf::from(dep).file_name().unwrap())
+                    .with_extension("stamp");
+
+                let output_paths = run_cargo(builder, cargo, vec![], &stamp, vec![], false, false);
+                let directories = output_paths
+                    .into_iter()
+                    .filter_map(|p| p.parent().map(ToOwned::to_owned))
+                    .fold(HashSet::new(), |mut set, dir| {
+                        set.insert(dir);
+                        set
+                    });
+                lib_paths.extend(directories);
+            }
+            lib_paths
+        } else {
+            vec![]
+        };
```

Comment:
The original idea @ehuss had was to put `trpl` (and theoretically any other such deps) in `rustbook`’s dependencies. Combined with *this*, that produced two copies of `libtrpl-<hash>.rlib` (and, I think, the dependencies for `trpl`). In principle, that might be the *only* thing we need to do here, Calling `builder.tool_exe()` via `builder.tool_cmd(tool:Rustbook)` guarantees that all of the relevant dependencies *are* built. Then we don’t need to do anything except determine what the appropriate path is—basically, the *logic* here, but none of the actual Cargo invocations?

I am going to test that locally, but I also wanted to put it here to get input from people who actually know the intent!

Reply:
Update here for posterity: this actually needed to go the *other* direction. Solely adding the dep to `rustbook` ended up failing to have all of the dependencies of `trpl` present when we actually needed to do the build. (I haven’t dug deeply to see *why* that is, but it has been the case consistently.) Instead, we keep this logic and do *not* add it to the dependencies for `rustbook` at all.